src/lib/certmap/sss_cert_digest_crypto.c

```diff
@@ -0,0 +1,169 @@
+/*
+   SSSD - digest functions - OpenSSL version
+   The calls defined here should be usable outside of SSSD as well, e.g. in
+   libsss_certmap.
+
+   Copyright (C) Sumit Bose <sbose@redhat.com> 2022
+
+   This program is free software; you can redistribute it and/or modify
+   it under the terms of the GNU General Public License as published by
+   the Free Software Foundation; either version 3 of the License, or
+   (at your option) any later version.
+
+   This program is distributed in the hope that it will be useful,
+   but WITHOUT ANY WARRANTY; without even the implied warranty of
+   MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+   GNU General Public License for more details.
+
+   You should have received a copy of the GNU General Public License
+   along with this program.  If not, see <http://www.gnu.org/licenses/>.
+*/
+
+#include "config.h"
+
+#include <string.h>
+#include <ctype.h>
+#include <errno.h>
+#include <stdbool.h>
+#include <talloc.h>
+#include <openssl/opensslv.h>
+#include <openssl/objects.h>
+#include <openssl/evp.h>
+
+#include "lib/certmap/sss_certmap_int.h"
+
+#if OPENSSL_VERSION_NUMBER < 0x30000000L
+#define discard_const(ptr) ((void *)((uintptr_t)(ptr)))
+#define EVP_MD_free(ctx)
+#define EVP_MD_fetch(ctx, algorithm, properties) discard_const(EVP_get_digestbyname(algorithm))
+#endif
+
+#define DATA_STEP_SIZE 30
+
+struct get_digest_data {
+    const char **list;
+    size_t size;
+    size_t idx;
+    int error;
+};
+
+static void get_digest_helper(const OBJ_NAME *name, void *arg)
+{
+    struct get_digest_data *data = (struct get_digest_data *) arg;
+    EVP_MD *md = NULL;
+
+    if (data->error != 0) {
+        return;
+    }
+
+    if (name == NULL || name->name == NULL || islower(*name->name) == 0
+        || strstr(name->name, "RSA") != NULL
+        || strstr(name->name, "rsa") != NULL) {
```

Comment:
What is the reason behind `islower()` and non RSA/rsa checks?

Reply:
I added a comment in the code, basically there some functions in the global digest list which are not plain hashes but signature functions.

Reply:
Check for 'RSA' is redundant after `islower()` check.

Reply:
`islower()` is only checking the first character, `RSA` can be present inside a name as well, e.g. `sha256WithRSAEncryption`.